Switch from Travis CI to GitHub Actions

krobelus opened the pull request:
Closes #1061

Simplistic port of the old Travis configuration. The results looks like this: [ubuntu.yml](https://github.com/krobelus/tig/actions/runs/468621862), [macos.yml](https://github.com/krobelus/tig/actions/runs/468621859).

This runs all eight combinations of compilers and build flags. We could select an interesting subset. Additionally we could run tests on other OS.

koutcher commented:
This is really great! I think we should move Cygwin builds from AppVeyor to GitHub Actions as well.
The full matrix could be a bit overkill (and time consuming). Previously address-sanitizer was only run with clang and valgrind with gcc.
Badges will have to be updated in README.adoc.

krobelus commented:
Ok I've updated the badges, and made the matrix more selective, so only 6 configurations are run instead of 8 on ubuntu.
I also added a Windows build but it fails to find <sys/wait.h>, which should be included in the Cygwin package [cygwin-devel](https://cygwin.com/cgi-bin2/package-cat.cgi?file=x86_64%2Fcygwin-devel%2Fcygwin-devel-3.1.7-1&grep=sys%2Fwait).
I didn't set up cygwin, so debugging is somewhat difficult. I don't think GitHub Actions provide SSH access.

koutcher commented on README.adoc:
```diff
@@ -2,8 +2,9 @@ Tig: text-mode interface for Git
-image:https://ci.appveyor.com/api/projects/status/jxt1uf52o7r0a8r7/branch/master?svg=true[AppVeyor Build,link=https://ci.appveyor.com/project/fonseca/tig]
+image:https://github.com/jonas/tig/workflows/Linux/badge.svg[Linux CI,link=https://github.com/jonas/tig/actions?query=workflow%3A.github%2Fworkflows%2Flinux.yml]
+image:https://github.com/jonas/tig/workflows/MacOS/badge.svg[MacOS CI,link=https://github.com/jonas/tig/actions?query=workflow%3A.github%2Fworkflows%2Fmacos.yml]
+image:https://github.com/jonas/tig/workflows/Windows/badge.svg[Ubuntu CI,link=https://github.com/jonas/tig/actions?query=workflow%3A.github%2Fworkflows%2Fwindows.yml]
```
Typo: Ubuntu -> Windows. I believe Mac OS should also now be referred as macOS.

koutcher commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,25 @@
+          packages: git gcc-code make automake libiconv-devel libncurses-devel cygwin-devel glibc
+      - name: Test Tig
+        env:
+          CC: gcc
```
Maybe setting the PATH environment variable could make a difference as appveyor.yml has a line with ` - cmd: SET PATH=C:\%TARGET%\bin;%PATH%`.

krobelus replied:
Right, we actually need to prepend to the path to use the right binaries. I have made this work on a Windows system now. The tests work fine (but are really slow). I don't know a workaround for https://github.com/actions/runner/issues/241 that works on Windows, so I commented out `make test`. We can probably just leave it like that for now.

koutcher replied:
Nice job. As we run the full test on Linux, it seems to be enough to just test the build on macOS and Windows. We can wait a feedback from @jonas to see if appveyor.yml can go as well.

krobelus replied:
The Windows job is not very interesting because with Cygwin it's using the same GNU toolchain as Linux.
It is, however useful as documentation of how to quickly get Tig up and running on Windows. I'm no longer using the GitHub Actions library that abstract installing cygwin, because a single `choco install cygwin` is enough; that makes it easier to recreate locally as well.

Maybe I'll try to run tests on macOS as well (provided the `script` workaround to get a TTY works there).

Could also be interesting to test on different libcs and other LInux distribution (which would also document which packages to install). But that's all not super important I guess.

koutcher replied:
We have seen in the past some Windows specific compilation issues (#858), so I think it is worth having the Windows job.
`script` should work on macOS but I don't know how we can run the tests with the Brew build.

koutcher replied:
Adding this step to macos.yml should make Homebrew run the test suite for us:
```
    - name: Patch
      run: |
        cd "$(brew --repo homebrew/core)" && patch -p1 <<\EOF
diff --git a/Formula/tig.rb b/Formula/tig.rb
index 2e8c125660..f76d6bca1c 100644
--- a/Formula/tig.rb
+++ b/Formula/tig.rb
@@ -26,9 +26,11 @@ class Tig < Formula
   depends_on "readline"
 
   def install
+    ENV.deparallelize
     system "./autogen.sh" if build.head?
     system "./configure", "--prefix=#{prefix}", "--sysconfdir=#{etc}"
     system "make"
+    system "make", "test"
     # Ensure the configured `sysconfdir` is used during runtime by
     # installing in a separate step.
     system "make", "install", "sysconfdir=#{pkgshare}/examples"
EOF
```

krobelus replied:
yep, that works

koutcher commented on .github/workflows/macos.yml:
```diff
@@ -0,0 +1,17 @@
+
+    steps:
+      - name: Install latest Tig
+        run: CC=${{ matrix.compiler }} HOMEBREW_NO_AUTO_UPDATE=1 brew install --HEAD tig
```
I don't think that Homebrew will honour the CC variable here. `HOMEBREW_CC=${{ matrix.compiler }}` may be a better choice.

krobelus replied:
makes sense, done

koutcher commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,25 @@
+          make configure
+          sh configure
+          make
+          # make test # TODO There is no TTY on Windows, see https://github.com/actions/runner/issues/241
```
Have you tried running `make test` on Windows ? When running in CMD.EXE, there is no TTY anyway, so the script hack may not be necessary for the Cygwin build...

krobelus replied:
yeah, it [fails](https://github.com/krobelus/tig/runs/1676952058?check_suite_focus=true) due to the missing TTY

On a real Windows machine the tests work (but run extremely slowly).

koutcher replied:
Ok then I believe we should drop windows.yml for now and keep relying on the AppVeyor build for Windows CI until there is a solution for this problem. Don't forget to put the AppVeyor badge back in README.adoc and I think we are good to go.

krobelus replied:
ok, done
